cpp/tests/streams/copying_test.cpp

```diff
@@ -0,0 +1,346 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cudf/utilities/default_stream.hpp"
+#include "cudf/utilities/type_dispatcher.hpp"
+#include "cudf_test/iterator_utilities.hpp"
+#include <cudf/copying.hpp>
+#include <cudf/detail/null_mask.hpp>
+
+#include <cudf_test/base_fixture.hpp>
+#include <cudf_test/column_wrapper.hpp>
+#include <cudf_test/default_stream.hpp>
+#include <cudf_test/type_lists.hpp>
+
+#include <limits>
+#include <utility>
+
+class CopyingTest : public cudf::test::BaseFixture {};
+
+TEST_F(CopyingTest, Gather)
+{
+  constexpr cudf::size_type source_size{1000};
+
+  auto data = cudf::detail::make_counting_transform_iterator(0, [](auto i) { return i; });
+  cudf::test::fixed_width_column_wrapper<int32_t> source_column(data, data + source_size);
+  cudf::test::fixed_width_column_wrapper<int32_t> gather_map(data, data + source_size);
+
+  cudf::table_view source_table({source_column});
+
+  std::unique_ptr<cudf::table> result = cudf::gather(source_table,
```

Comment:
Good question! There are no assertions in these tests because we aren't asserting any invariants. What we are doing is calling an API (e.g. `cudf::gather`) and then relying on some hidden machinery to verify that the same stream is actually used in every CUDA API call that is performed inside `cudf::gather`. See https://github.com/rapidsai/cudf/pull/13629#discussion_r1260335119 for more info (and a useful link).

Regarding the unused outputs, that's a good call. It's probably not causing any issues because I'm guessing we don't compile tests with warnings as errors, only libcudf itself. That said, it would be good to remove these. I'll fix.